compiler/haskell-ide-core/src/Development/IDE/Types/Diagnostics.hs

```diff
@@ -3,137 +3,230 @@
 import Data.Text.Prettyprint.Doc.Syntax
-import GHC.Generics
-import qualified Network.URI.Encode
+import Data.String (IsString(..))
 import qualified Text.PrettyPrint.Annotated.HughesPJClass as Pretty
+import           Language.Haskell.LSP.Types as LSP (
+    DiagnosticSeverity(..)
+  , Diagnostic(..)
+  , filePathToUri
+  , uriToFilePath
+  , List(..)
+  , DiagnosticRelatedInformation(..)
+  , Uri(..)
+  )
+import Language.Haskell.LSP.Diagnostics
 
 import Development.IDE.Types.Location
 
-ideErrorText :: FilePath -> T.Text -> Diagnostic
-ideErrorText absFile = errorDiag absFile "Ide Error"
-
-ideErrorPretty :: Pretty.Pretty e => FilePath -> e -> Diagnostic
-ideErrorPretty absFile = ideErrorText absFile . T.pack . Pretty.prettyShow
-
-errorDiag :: FilePath -> T.Text -> T.Text -> Diagnostic
-errorDiag fp src msg =
-  Diagnostic
-  { dFilePath = fp
-  , dRange    = noRange
-  , dSeverity = Error
-  , dSource   = src
-  , dMessage  = msg
+ideErrorText :: FilePath -> T.Text -> LSP.Diagnostic
+ideErrorText fp = errorDiag fp "Ide Error"
+
+ideErrorPretty :: Pretty.Pretty e => FilePath -> e -> LSP.Diagnostic
+ideErrorPretty fp = ideErrorText fp . T.pack . Pretty.prettyShow
+
+errorDiag :: FilePath -> T.Text -> T.Text -> LSP.Diagnostic
+errorDiag fp src =
+  set dFilePath (Just fp) . diagnostic noRange LSP.DsError src
+
+-- | This is for compatibility with our old diagnostic type
+diagnostic :: Range
+           -> LSP.DiagnosticSeverity
+           -> T.Text -- ^ source
+           -> T.Text -- ^ message
+           -> LSP.Diagnostic
+diagnostic rng sev src msg
+    = LSP.Diagnostic {
+          _range = rng,
+          _severity = Just sev,
+          _code = Nothing,
+          _source = Just src,
+          _message = msg,
+          _relatedInformation = Nothing
+          }
+
+-- | Any optional field is instantiated to Nothing
+defDiagnostic ::
+  Range ->
+  T.Text -> -- ^ error message
+  LSP.Diagnostic
+defDiagnostic _range _message = LSP.Diagnostic {
+    _range
+  , _message
+  , _severity = Nothing
+  , _code = Nothing
+  , _source = Nothing
+  , _relatedInformation = Nothing
   }
 
-ideTryIOException :: FilePath -> IO a -> IO (Either Diagnostic a)
+-- | setLocation but with no range information
+dFilePath ::
+  Lens' LSP.Diagnostic (Maybe FilePath)
+dFilePath = lens g s where
+    g :: LSP.Diagnostic -> Maybe FilePath
+    g d = (uriToFilePath . _uri) =<< view dLocation d
+    s :: LSP.Diagnostic -> (Maybe FilePath) -> LSP.Diagnostic
+    s d fp = set dLocation (Location <$> (filePathToUri <$> fp) <*> pure noRange) d
+
+-- | This adds location information to the diagnostics but this is only used in
+--   the case of serious errors to give some context to what went wrong
+dLocation ::
+  Lens' LSP.Diagnostic (Maybe Location)
+dLocation = lens g s where
+    s :: LSP.Diagnostic -> Maybe Location -> LSP.Diagnostic
+    s d = \case
+        Just loc -> d {LSP._relatedInformation = Just $ LSP.List [DiagnosticRelatedInformation loc ""]}
```

Comment:
Should we add "dLocation: Unknown error" here just so we ever see blank errors we know why they are showing up?

Reply:
Good call